#####Stream lining the analysis of BUSTED and BUSTED+SRV data
# All of this is written assuming you only have 3 alpha and omega classes.
# adjustments can be made for otherwise but it'll take a minute
# also tends to assume you have gene alignmets in all 4 sig cats
# so if you don't have any in one cat it might throw an error
#further investigation needed

setwd("~/Google Drive/Muse/2-4-2016/Files for 2-24/")
srv = read.delim(file="C:/Users/srwisots/Drive/Muse/2-4-2016/Files for 2-24/srv-final.txt", header=TRUE,sep="\t",as.is = T);
attach(srv)

setwd("C:/Users/srwisots/Drive/Muse/Simulation for SRv/large.nex Sim/results")
test1 = read.delim(file="test1.txt", header = TRUE, sep = "\t", as.is = T)
test2 = read.delim(file="test2.txt", header = TRUE, sep = "\t", as.is = T)

###process data
# function that adds significant categories based on p values
# and correctly orders and normalizes the alphas if redo.alphas = TRUE
process.dat <- function(dat, redo.alphas = TRUE){
  ### Add sig categories
  dat= cbind(dat,p.cat=cut(dat$BUSTED.P,breaks=c(-Inf,0.05,Inf),
                           labels = c("S V B", "NS V B")))
  dat= cbind(dat,p.srv.cat=cut(dat$BUSTED.SRV.P,breaks=c(-Inf,0.05,Inf),
                               labels = c("S V BSRV", "NS V BSRV")))
  ###rearrange alphas
  if(redo.alphas==TRUE){
    normalize = dat$SRV.alpha1.MLE*dat$SRV.alpha1.prop + 
      dat$SRV.alpha2.MLE*dat$SRV.alpha2.prop+ 
      dat$SRV.alpha3.MLE*dat$SRV.alpha3.prop
    
    
    dat = cbind(dat,true.alpha1 = dat$SRV.alpha1.MLE/normalize,
                true.alpha2 = dat$SRV.alpha2.MLE/normalize,
                true.alpha3 = dat$SRV.alpha3.MLE/normalize)
  }
  else{
    #sets the true.alpha values to just the standard MLE 
    #makes the rest of the stream line analysis easier. 
    #it's either this or make you specify alpha variables in later steps...
    dat = cbind(dat,true.alpha1 = dat$SRV.alpha1.MLE,
                true.alpha2 = dat$SRV.alpha2.MLE,
                true.alpha3 = dat$SRV.alpha3.MLE)  }
}



####Create Box plots


#### Summary Stats ------
# here's where you'd have to start specifying other alphas 
# returns a list of statistics generated from the boxplot function 
# note the omega3 value here is from BUSTED SRV 
# not the omega3 from BUSTED for the same file
box.sum.stats <- function(dat, subset = NULL){
  omega3 = boxplot(
    dat$BUSTED.SRV.omega3.MLE ~ dat$p.cat + dat$p.srv.cat, data = dat, 
    subset = subset, plot = FALSE
  )
  
  alpha1 = boxplot(
    dat$true.alpha1 ~ dat$p.cat + dat$p.srv.cat, data = dat,
    subset = subset, plot = FALSE
  )
  
  alpha2 = boxplot(
    dat$true.alpha2 ~ dat$p.cat + dat$p.srv.cat, data = dat,
    subset = subset, plot = FALSE
  )
  
  alpha3 = boxplot(
    dat$true.alpha3 ~ dat$p.cat + dat$p.srv.cat, data = dat,
    subset = subset,plot = FALSE
  )
  results <- list(omega3 =omega3, alpha1 = alpha1,alpha2= alpha2, alpha3 = alpha3)
  return(results)
}

#### Generate a ton of boxplots ------
#dat is the BUSTED+SRV data
#subset is the subset of that data you which to investigate
#box.stats are the boxplot statistics generated by box.sum.stats
#these are used to set the axis limits
#only works if there are files in all 4 categories
gen.boxplots <- function(dat, subset =NULL){
  ####alpha1  ---------------
  png("alpha1.png")
  boxplot(
    dat$true.alpha1 ~ dat$p.cat + dat$p.srv.cat, data = dat,
    subset = subset, 
    horizontal = T, ylab = "P-value category", 
    xlab = "Alpha1 & Omega3 for BUSTED+SRV",
    names = c("","","",""),
    col = rgb(0,0,1,0.5)
  )
  
  boxplot(
    dat$BUSTED.SRV.omega3.MLE ~ dat$p.cat + dat$p.srv.cat, data = dat,
    subset = subset,
    horizontal = T, ylab = "P-value category",
    names = c("","","",""),
    col = rgb(1,0,0,0.5), add = T
  )
  axis(2,at=c(1,2,3,4),labels=c("Both","Only \n BUSTED+SRV","Only \nBUSTED","No \n Selection"), cex.axis = 0.6)
  legend("topright",c("Alpha1", "Omega3"), fill = c(rgb(0,0,1,0.5),rgb(1,0,0,0.5) ))
  dev.off()
  ##### alpha2 -----------
  png("alpha2.png")
  boxplot(
    dat$true.alpha2 ~ dat$p.cat + dat$p.srv.cat, data = dat, 
    subset = subset,
    ylim = c(0,max(box.stats$omega3$stats)),
    horizontal = T, ylab = "P-value category", 
    xlab = "Alpha2 & Omega3 for BUSTED+SRV", las =1,
    names = c("","","",""),
    col = rgb(0,0,1,0.5)
  )
  boxplot(
    dat$BUSTED.SRV.omega3.MLE ~ dat$p.cat + dat$p.srv.cat, data = dat, 
    subset = subset,
    horizontal = T, ylab = "P-value category", las = 1,
    names = c("","","",""),
    col = rgb(1,0,0,0.5), add = T
  )
  axis(2,at=c(1,2,3,4),labels=c("Both","Only \n BUSTED+SRV","Only \nBUSTED","No \n Selection"), cex.axis = 0.6)
  legend("topright",c("alpha2", "Omega3"), fill = c(rgb(0,0,1,0.5),rgb(1,0,0,0.5) ))
  dev.off()
  ##### alpha 3 ------
  png("alpha3.png")
  boxplot(
    dat$true.alpha3 ~ dat$p.cat + dat$p.srv.cat, data = dat, 
    subset = subset,
    ylim = c(0,max(box.stats$omega3$stats)),
    horizontal = T, ylab = "P-value category", 
    xlab = "alpha3 & Omega3 for BUSTED+SRV", las = 1,
    names = c("","","",""), 
    col = rgb(0,0,1,0.5)
  )
  boxplot(
    dat$BUSTED.SRV.omega3.MLE ~ dat$p.cat + dat$p.srv.cat, data = dat, 
    subset = subset,
    horizontal = T, ylab = "P-value category", las = 1,
    names = c("","","",""), col = rgb(1,0,0,0.5), add = T
  )
  axis(2,at=c(1,2,3,4),labels=c("Both","Only \n BUSTED+SRV","Only \nBUSTED","No \n Selection"), cex.axis = 0.6)
  legend("topright",c("Alpha3", "Omega3"), fill = c(rgb(0,0,1,0.5),rgb(1,0,0,0.5) ))
  dev.off()
  
  ###alpha comparison boxplots
  png("alpha comparisons.png")
  boxplot(
    dat$true.alpha3 ~ dat$p.cat + dat$p.srv.cat, data = dat, 
    subset = subset,
    ylim = c(0,8),
    horizontal = T, ylab = "P-value category", xlab = "alpha1, alpha2, alpha3 ", las = 1,
    names = c(names = c("","","","")), col = rgb(0,0,1,0.5)
  )
  
  boxplot(
    dat$true.alpha2 ~ dat$p.cat + dat$p.srv.cat, data = dat,
    subset = subset, 
    horizontal = T, ylab = "P-value category", 
    names =  c("","","",""), col = rgb(1,0,1,0.5), add = T
  )
  
  boxplot(
    dat$true.alpha1 ~ dat$p.cat + dat$p.srv.cat, data = dat,
    subset =subset, 
    horizontal = T, ylab = "P-value category", 
    names =  c("","","",""), col = rgb(0,1,1,0.5), add =T
  )
  axis(2,at=c(1,2,3,4),labels=c("Both","Only \n BUSTED+SRV","Only \nBUSTED","No \n Selection"), cex.axis = 0.6)
  legend("topright",c("alpha3", "alpha2", "alpha1"), 
         fill = c(rgb(0,0,1,0.5),rgb(1,0,1,0.5),rgb(0,1,1,0.5)))
  dev.off()
  
}

##### Other graphs -------
#just basic BUSTED.P vs BUSTED.SRV.P plot
gen.p.plot <- function(dat,file){
  png(file)
  par(mfrow= c(2,2))
  plot(dat$BUSTED.P,dat$BUSTED.SRV.P)
  abline(a=0,b=1, col = "red")
  plot(dat$BUSTED.P,dat$BUSTED.SRV.P,xlim=c(0,0.5),ylim=c(0,0.5))
  abline(a=0,b=1,col="red")
  plot(dat$BUSTED.P,dat$BUSTED.SRV.P,xlim=c(0,0.25),ylim=c(0,0.25))
  abline(a=0,b=1,col="red")
  plot(dat$BUSTED.P,dat$BUSTED.SRV.P,xlim=c(0,0.1),ylim=c(0,0.1))
  abline(a=0,b=1,col="red")
  par(mfrow=c(1,1))
  dev.off()
}
install.packages("xtable")
#generate html table that seperates the gene alignments into the 4 sig categories
gen.sig.table <- function(dat){
  require("xtable")
  library("xtable")
  
  under.sel.busted = which(dat$BUSTED.P<=0.05)
  under.sel.srv = which(dat$BUSTED.SRV.P<=0.05)
  if( length(under.sel.srv) == 0){
    
    mat = matrix( rep(0),nrow=length(dat$BUSTED.P),ncol = 2,
                  dimnames= list(1:length(dat$BUSTED.P),
                                 c("BUSTED", "BUSTED-SRV")))
    mat[under.sel.busted,1] = 1
    mat[under.sel.srv,2] = 1 
    sel.tab = table(mat[,1],mat[,2], dnn = colnames(mat))
    row.names(sel.tab) = c("Selection")
    colnames(sel.tab) = c("No Selection")
    Sel.Prop =prop.table(sel.tab)
    test.table = xtable(Sel.Prop)
    print.xtable(test.table, type = "html", file="sel-table.html")
    return("All alignments are not significant according to BUSTED+SRV")
  }
  else if( length(under.sel.busted) == 0){
    
    mat = matrix( rep(0),nrow=length(dat$BUSTED.P),ncol = 2,
                  dimnames= list(1:length(dat$BUSTED.P),
                                 c("BUSTED", "BUSTED-SRV")))
    mat[under.sel.busted,1] = 1
    mat[under.sel.srv,2] = 1 
    sel.tab = table(mat[,1],mat[,2], dnn = colnames(mat))
    row.names(sel.tab) = c("No Selection")
    colnames(sel.tab) = c("Selection")
    Sel.Prop =prop.table(sel.tab)
    test.table = xtable(Sel.Prop)
    print.xtable(test.table, type = "html", file="sel-table.html")
    return("All alignments are not significant according to BUSTED")
  }
  else {
    mat = matrix( rep(0),nrow=length(dat$BUSTED.P),ncol = 2,
                  dimnames= list(1:length(dat$BUSTED.P),
                                 c("BUSTED", "BUSTED-SRV")))
    mat[under.sel.busted,1] = 1
    mat[under.sel.srv,2] = 1 
    sel.tab = table(mat[,1],mat[,2], dnn = colnames(mat))
    row.names(sel.tab) = c("No Selection", "Selection")
    colnames(sel.tab) = c("No Selection", "Selection")
    Sel.Prop =prop.table(sel.tab)
    test.table = xtable(Sel.Prop)
    print.xtable(test.table, type = "html", file="sel-table.html")
  }
}

# generate fraction under selection graphs
# only works out for if you have gene alignments in all 4 rate categories
gen.frac.sel.graphs <- function(dat){
  library("Hmisc")
  
  #frac under selection vs tree length
  
  #create bins of at least 100 data sets based on tree length
  
  dat = cbind(dat, tree.srv.cat = cut2(
    dat$BUSTED.SRV.treelength, m = 100, levels.mean = TRUE
  ))
  
  #data = cbind(data, tree.cat = cut2(data$BUSTED.treelength, m = 100, levels.mean = TRUE))
  #then find the fraction of data sets under positive selection according to
  #BUSTED+SRV p <= 0.05 in each bin
  
  frac.tree.srv = with(dat, tapply(BUSTED.SRV.P, tree.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  frac.tree = with(dat, tapply(BUSTED.P, tree.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  
  
  #frac under selection vs omega 3
  
  dat = cbind(dat, omega3.srv.cat = cut2(
    dat$BUSTED.SRV.omega3.MLE  , m = 100, levels.mean = TRUE
  ))
  
  #data = cbind(data, omega3.cat = cut2(data$BUSTED.omega3.MLE  , m = 100, levels.mean = TRUE))
  #then find the fraction of data sets under positive selection according to
  #BUSTED+SRV p <= 0.05 in each bin
  
  frac.omega3.srv = with(dat, tapply(BUSTED.SRV.P, omega3.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  frac.omega3 = with(dat, tapply(BUSTED.P, omega3.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  
  
  #frac under selection vs codons
  
  #create bins of at least 100 data sets based on tree length
  
  dat = cbind(dat, sites.srv.cat = cut2(dat$Sites, m = 100, levels.mean = TRUE))
  
  #data = cbind(data, sites.cat = cut2(data$Sites, m = 100, levels.mean = TRUE))
  #then find the fraction of data sets under positive selection according to
  #BUSTED+SRV p <= 0.05 in each bin
  
  frac.codons.srv = with(dat, tapply(BUSTED.SRV.P, sites.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  frac.codons = with(dat, tapply(BUSTED.P, sites.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  
  
  
  #frac under selection vs codons
  
  
  #create bins of at least 100 data sets based on tree length
  
  dat = cbind(dat, Sequences.cat = cut2(dat$Sequences, m = 100, levels.mean = TRUE))
  
  #then find the fraction of dat sets under positive selection according to
  #BUSTED+SRV p <= 0.05 in each bin
  
  frac.seq.srv = with(dat, tapply(dat$BUSTED.SRV.P, Sequences.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  frac.seq = with(dat, tapply(dat$BUSTED.P, Sequences.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  
  
  ##### PLOTS ------------------------------
  pdf(file = "Fraction Under Selection-Binned by SRV.pdf")
  par(mfrow = c(2,2))
  
  ## CODONS
  plot(
    x = as.numeric(levels(dat$sites.srv.cat)),y = frac.codons.srv, ylim = c(0,1), pch = 19,
    xlab = "Codons", ylab = "Fraction under selection"
  )
  points(
    x = as.numeric(levels(dat$sites.srv.cat)),y = frac.codons, col = 'red', pch = 19
  )
  lines(lowess(x = as.numeric(levels(dat$sites.srv.cat)), y = frac.codons.srv, f = 0.25))
  lines(lowess(x = as.numeric(levels(dat$sites.srv.cat)), y = frac.codons, f = 0.25))
  
  ## Sequences
  
  plot(
    x = as.numeric(levels(dat$Sequences.cat)), y = frac.seq.srv, ylim = c(0,1), pch = 19,
    xlab = "Sequences", ylab = "Fraction under selection"
  )
  points(
    x = as.numeric(levels(dat$Sequences.cat)), y = frac.seq, col = 'red', pch = 19
  )
  lines(lowess(x = as.numeric(levels(dat$Sequences.cat)), y = frac.seq.srv, f = 0.25))
  lines(lowess(x = as.numeric(levels(dat$Sequences.cat)), y = frac.seq, f = 0.25))
  
  ## TREE LENGTH
  
  plot(
    x = as.numeric(levels(dat$tree.srv.cat))[1:65], y = frac.tree.srv[1:65],
    ylim = c(0,1), pch = 19,
    xlab = "Tree length", ylab = "Fraction under selection"
  )
  points(
    x = as.numeric(levels(dat$tree.srv.cat))[1:65],y = frac.tree[1:65], col = 'red', pch = 19
  )
  lines(lowess(x = as.numeric(levels(dat$tree.srv.cat))[1:65], y = frac.tree.srv[1:65], f = 0.25))
  lines(lowess(x = as.numeric(levels(dat$tree.srv.cat))[1:65],y = frac.tree[1:65], f = 0.25))
  ## OMEGA 3
  
  plot(
    x = as.numeric(levels(dat$omega3.srv.cat))[1:52],y = frac.omega3.srv[1:52], ylim = c(0,1),
    xlim = c(0,60),pch = 19,
    xlab = "Omega3", ylab = "Fraction under selection"
  )
  points(
    x = as.numeric(levels(dat$omega3.srv.cat))[1:52], y = frac.omega3[1:52], col = 'red', pch = 19
  )
  lines(lowess(x = as.numeric(levels(
    dat$omega3.srv.cat
  ))[1:52],y = frac.omega3.srv[1:52], f = 0.25))
  lines(lowess(x = as.numeric(levels(dat$omega3.srv.cat))[1:52], y = frac.omega3[1:52], f = 0.25))
  dev.off()
  
  
  # Omega Prop --------------------------------------------------------------
  
  # Omega3
  # data = cbind(data, omega3.srv.cat = cut2(data$BUSTED.SRV.omega3.MLE  , m = 100))
  # data = cbind(data, omega3.cat = cut2(data$BUSTED.omega3.MLE  , m = 100, levels.mean = TRUE))
  # #then find the fraction of data sets under positive selection according to
  # #BUSTED+SRV p <= 0.05 in each bin
  # 
  # frac.omega3.srv = with(data, tapply(BUSTED.SRV.P, omega3.srv.cat, function(x)
  #   length(which(x <= 0.05)) / length(x)))
  
  ###### OMEGA 1
  
  dat = cbind(dat, omega1.srv.cat = cut2(
    dat$BUSTED.SRV.omega1.MLE  , m = 100, levels.mean = TRUE
  ))
  # dat = cbind(data, omega1.cat = cut2(data$BUSTED.omega1.MLE  , m = 100, levels.mean = TRUE))
  #then find the fraction of data sets under positive selection according to
  #BUSTED+SRV p <= 0.05 in each bin
  
  frac.omega1.srv = with(dat, tapply(BUSTED.SRV.P, omega1.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  frac.omega1 = with(dat, tapply(BUSTED.P, omega1.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  
  ##### OMEGA 2
  
  dat = cbind(dat, omega2.srv.cat = cut2(
    dat$BUSTED.SRV.omega2.MLE  , m = 100, levels.mean = TRUE
  ))
  
  # data = cbind(data, omega2.cat = cut2(data$BUSTED.omega2.MLE  , m = 100, levels.mean = TRUE))
  #then find the fraction of data sets under positive selection according to
  #BUSTED+SRV p <= 0.05 in each bin
  
  frac.omega2.srv = with(dat, tapply(BUSTED.SRV.P, omega2.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  frac.omega2 = with(dat, tapply(BUSTED.P, omega2.srv.cat, function(x)
    length(which(x <= 0.05)) / length(x)))
  
  pdf(file = "Omegas-binned by SRV.pdf")
  par(mfrow = c(3,1))
  
  #### OMEGA 1
  
  plot(
    x = as.numeric(levels(dat$omega1.srv.cat))[1:58],y = frac.omega1.srv[1:58], ylim = c(0,1),
    pch = 19,
    xlab = "omega1", ylab = "Fraction under selection"
  )
  points(
    x = as.numeric(levels(dat$omega1.srv.cat))[1:58], y = frac.omega1[1:58], col = 'red', pch = 19
  )
  lines(lowess(x = as.numeric(levels(
    dat$omega1.srv.cat
  ))[1:58],y = frac.omega1.srv[1:58], f = 0.25))
  lines(lowess(x = as.numeric(levels(dat$omega1.srv.cat))[1:58], y = frac.omega1[1:58], f = 0.25))
  
  #### PLOT OMEGA 2
  
  plot(
    x = as.numeric(levels(dat$omega2.srv.cat)),y = frac.omega2.srv, ylim = c(0,1),
    pch = 19,
    xlab = "omega2", ylab = "Fraction under selection"
  )
  points(
    x = as.numeric(levels(dat$omega2.srv.cat)), y = frac.omega2, col = 'red', pch = 19
  )
  lines(lowess(x = as.numeric(levels(
    dat$omega2.srv.cat
  )),y = frac.omega2.srv, f = 0.25))
  lines(lowess(x = as.numeric(levels(dat$omega2.srv.cat)), y = frac.omega2, f = 0.25))
  
  #### PLOT OMEGA 3
  
  plot(
    x = as.numeric(levels(dat$omega3.srv.cat))[1:52],y = frac.omega3.srv[1:52], ylim = c(0,1),
    xlim = c(0,60),pch = 19,
    xlab = "Omega3", ylab = "Fraction under selection"
  )
  points(
    x = as.numeric(levels(dat$omega3.srv.cat))[1:52], y = frac.omega3[1:52], col = 'red', pch = 19
  )
  lines(lowess(x = as.numeric(levels(
    dat$omega3.srv.cat
  ))[1:52],y = frac.omega3.srv[1:52], f = 0.25))
  lines(lowess(x = as.numeric(levels(dat$omega3.srv.cat))[1:52], y = frac.omega3[1:52], f = 0.25))
  dev.off()
}

#run time graph
run.time <- function(dat){
  png("runtime.png")
  plot(x = dat$BUSTED.Run.Time, y=dat$BUSTED.SRV.Run.Time,
       xlab = "Run Time for \n Unconstrained BUSTED",
       ylab = 'Run Time for \n Unconstrained BUSTED+SRV')
  abline(a= 0, b =1, col = 'red')
  dev.off()
  ratio = mean(dat$BUSTED.SRV.Run.Time)/mean(dat$BUSTED.Run.Time)
  return(ratio)
}
#generate pairs plots
#log axis by default 
# if you have 0's in p values log will throw an error
gen.pairs.plots <- function(dat, log = TRUE, name){
  if ( log == T){
    png(name, width = 8,height = 8, units = "in", res = 72)
    pairs(~BUSTED.P+BUSTED.SRV.P+BUSTED.omega3.MLE+BUSTED.SRV.omega3.MLE
          +true.alpha3, 
          data = dat, log ="xy")
    dev.off()
  }
  else {
    png(name, width = 8,height = 8, units = "in", res = 72) 
    pairs(~BUSTED.P+BUSTED.SRV.P+BUSTED.omega3.MLE+BUSTED.SRV.omega3.MLE+
            +true.alpha3, 
          data = dat)
    dev.off()
  }
}

do.it.all <- function(dat, name){
  dat = process.dat(dat)
  x = box.sum.stats(dat)
  gen.boxplots(dat, box.stats = x)
}

#checking Bias
plot(test1.process$BUSTED.omega3.MLE,test1.process$BUSTED.SRV.omega3.MLE
)
plot(test1.process$SRV.alpha3.MLE,
     ylim = c(0, 6.66))
abline(h= 6.66)
points(test1.process$SRV.alpha2.MLE, col = 2)
abline(h= 0.66, col = 2)
points(test1.process$SRV.alpha1.MLE, col = 3)
abline(h=0.133, col =3)

plot(test1.process$BUSTED.omega3.MLE)
abline(h=1)
plot(test1.process$BUSTED.omega2.MLE,col = 2)
abline(h = .5, col =2)
plot(test1.process$BUSTED.omega1.MLE,col = 3)
abline(h = .1, col =3)


plot(test1.process$BUSTED.SRV.omega3.MLE)
abline(h=1)
plot(test1.process$BUSTED.SRV.omega2.MLE,col = 2)
abline(h = .5, col =2)
plot(test1.process$BUSTED.SRV.omega1.MLE,col = 3)
abline(h = .1, col =3)
